### PR TITLE
expose user flags

### PR DESCRIPTION
**Pull request by JohnyTheCarrot**
### Summary

This pull request adds user flags using the public_flags field.
Library users can, using this, look at the hypesquad houses of a user, and their flags such as partner, staff, hypesquad etc.
This is my first time here, so please be gentle.

### Checklist

- [x] If code changes were made then they have been tested.
    - [ ] I have updated the documentation to reflect the changes.
- [ ] This PR fixes an issue.
- [x] This PR adds something new (e.g. new method or parameters).
- [ ] This PR is a breaking change (e.g. methods or parameters removed/renamed)
- [ ] This PR is **not** a code change (e.g. documentation, README, ...)

### CODE REVIEW

File: discord/member.py [hunk 1]
```diff
@@ -159,10 +159,12 @@ class Member(discord.abc.Messageable, _BaseUser):
     premium_since: Optional[:class:`datetime.datetime`]
         A datetime object that specifies the date and time in UTC when the member used their
         Nitro boost on the guild, if available. This could be ``None``.
+    user_flags: list[:class:`UserFlags`]
```
Comment (Rapptz) [inline discord/member.py:162]: This is actually not exposed anywhere. There is no `user_flags` attribute anywhere here.

File: discord/member.py [hunk 2]
```diff
@@ -171,6 +173,7 @@ def __init__(self, *, data, guild, state):
         self.joined_at = utils.parse_time(data.get('joined_at'))
         self.premium_since = utils.parse_time(data.get('premium_since'))
         self._update_roles(data)
+        self._public_flags = data['user'].get('public_flags', 0)
```
Comment (Rapptz) [inline discord/member.py:176]: If something is inside the `user` payload then surely this belongs within `User` and not `Member`.

**Comment by JohnyTheCarrot** [comment]
I think I've made the changes requested, let me know if there's anything else :)

File: discord/public_flags.py [hunk 3]
```diff
@@ -0,0 +1,96 @@
+from .enums import UserFlags
+
+
+class PublicFlags:
+    """Wraps up the Discord public flags.
+
+    .. versionadded :: 1.3.4
```
Comment (Rapptz) [inline discord/public_flags.py:7]:
```suggestion
    .. versionadded :: 1.4
```

Also these tags are generally added at the bottom of the operation table but above the attribute table. See `flags.py` for an example.

File: discord/public_flags.py [hunk 4]
```diff
@@ -0,0 +1,96 @@
+from .enums import UserFlags
+
+
+class PublicFlags:
```
Comment (Rapptz) [inline discord/public_flags.py:4]:
```suggestion
class PublicUserFlags:
```

This is actually not how flags are done in the library.

Please see `flags.py` for an example on how flags are done.

File: discord/user.py [hunk 5]
```diff
@@ -301,6 +308,8 @@ class ClientUser(BaseUser):
         Specifies if the user is a bot account.
     system: :class:`bool`
         Specifies if the user is a system user (i.e. represents Discord officially).
+    public_flags: :class:´PublicFlags´
```
Comment (avayert) [inline discord/user.py:311]:
```suggestion
    public_flags: :class:`PublicFlags`
```

File: discord/user.py [hunk 6]
```diff
@@ -82,7 +83,7 @@ def system(self):
 _BaseUser = discord.abc.User
 
 class BaseUser(_BaseUser):
-    __slots__ = ('name', 'id', 'discriminator', 'avatar', 'bot', 'system', '_state')
+    __slots__ = ('name', 'id', 'discriminator', 'avatar', 'bot', 'system', '_public_flags', '_state', 'public_flags')
```
Comment (Rapptz) [inline discord/user.py:86]:
```suggestion
    __slots__ = ('name', 'id', 'discriminator', 'avatar', 'bot', 'system', '_public_flags', '_state')
```

One of these isn't an actual attribute.

File: discord/user.py [hunk 7]
```diff
@@ -100,13 +101,19 @@ def __ne__(self, other):
     def __hash__(self):
         return self.id >> 22
 
+    def has_flag(self, flag: UserFlags):
+        v = flag.value
+        return (self._public_flags & v) == v
+
```
Comment (Rapptz) [inline discord/user.py:107]:
```suggestion
```

This method shouldn't be here -- it's part of the flag.

File: discord/user.py [hunk 8]
```diff
@@ -100,13 +101,19 @@ def __ne__(self, other):
...
     def _update(self, data):
         self.name = data['username']
         self.id = int(data['id'])
         self.discriminator = data['discriminator']
         self.avatar = data['avatar']
+        self._public_flags = data.get("public_flags", 0)
```
Comment (Rapptz) [inline discord/user.py:113]:
```suggestion
        self._public_flags = data.get('public_flags', 0)
```

Single quotes are used for strings here.

File: discord/user.py [hunk 9]
```diff
@@ -100,13 +101,19 @@ def __ne__(self, other):
...
         self.bot = data.get('bot', False)
         self.system = data.get('system', False)
+        self.public_flags = PublicFlags(self)
```
Comment (Rapptz) [inline discord/user.py:116]: This should be a property. See e.g. `Guild.system_message_flags` or `Message.flags` for an example of what I mean.

File: discord/user.py [hunk 10]
```diff
@@ -301,6 +308,8 @@ class ClientUser(BaseUser):
         Specifies if the user is a bot account.
     system: :class:`bool`
         Specifies if the user is a system user (i.e. represents Discord officially).
+    public_flags: :class:´PublicFlags´
+        The publicly available flags the user has.
```
Comment (Rapptz) [inline discord/user.py:312]: This documentation should just be in the property.

File: discord/user.py [hunk 11]
```diff
@@ -674,6 +683,8 @@ class User(BaseUser, discord.abc.Messageable):
         Specifies if the user is a bot account.
     system: :class:`bool`
         Specifies if the user is a system user (i.e. represents Discord officially).
+    public_flags: :class:`PublicFlags`
+        The publicly available flags the user has.
```
Comment (Rapptz) [inline discord/user.py:687]: This documentation should just be in the property.

File: discord/public_flags.py [hunk 12]
```diff
@@ -0,0 +1,96 @@
+from .enums import UserFlags
```
Comment (Rapptz) [inline discord/public_flags.py:1]: This is missing the license header present in every file, along with the encoding header.

**Comment by JohnyTheCarrot** [comment]
I think I've made all requested changes, if not, let me know if I forgot something/if you need anything else changed :)

**Comment by JohnyTheCarrot** [comment]
Sorry for the delay, I've been a bit busy. I'll get to it as soon as I can. :)

**Comment by JohnyTheCarrot** [comment]
Sorry for the massive delay. My code is ready for review. I assume the version needs to be updated?

**Pull request closed without merging**